### PR TITLE
Express notes api

### PR DESCRIPTION
**Pull request by Quintin-Russell**
image

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+// id does not exist = 404status+err
+app.get('/api/notes/:id', (req, res) => {
+  const id = req.params.id;
+  let object = null;
+  if (id < 0) {
+    const message = {
+      error: 'Please enter a valid ID'
+    };
+    res.status(400).json(message);
+  }
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:39]: What about other invalid `id` values, such as `0`, `13.52`, or `'jones'`?

---

image

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    cont = cont.content;
+    const entry = {
+      content: cont,
+      id: nextId
+    };
+    dataJSON.notes[nextId] = entry;
+    dataJSON.nextId = nextId + 1;
+    writeFile('data.json', dataJSON);
+    return res.status(201).json(entry);
+  } else {
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:71]: What happens if the request body contains the _wrong_ key name?

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    let cont = req.body;
+    cont = cont.content;
+    const entry = {
+      content: cont,
+      id: nextId
+    };
+    dataJSON.notes[nextId] = entry;
+    dataJSON.nextId = nextId + 1;
+    writeFile('data.json', dataJSON);
+    return res.status(201).json(entry);
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:70]: Successful responses to write operations should only be sent after the data file is successfully updated,

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]
@Quintin-Russell Please include screenshots validating each possible outcome of each route.

File: express-notes-api/index.js (hunk 1 above)
Comment (Quintin-Russell) [reply express-notes-api/index.js:39]: edited

File: express-notes-api/index.js (hunk 3 above)
Comment (Quintin-Russell) [reply express-notes-api/index.js:70]: edited

File: express-notes-api/index.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+});
+
+app.use(express.json());
+
+// issue 2: GET req w/ id returns the obj with that id; negative id = 400 status+err;
+// id does not exist = 404status+err
+app.get('/api/notes/:id', (req, res) => {
+  const id = req.params.id;
+  let object = null;
+  if ((id > 0) && ((id % 1) === 1)) {
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:34]: image

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]

File: express-notes-api/index.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+// issue 4: DELETE req:
+// xx- not a valid id(not a + integer): 400 status code + JSON error obj
+// xx- if +id but id doesnt exist: 404 status code + JSON error obj
+// xx- if valid id but error: next(err)
+// xxif valid id: delete note + 204 status
+
+app.delete('/api/notes/:id', (req, res) => {
+  const delId = req.params.id;
+  if ((delId > 0) && ((delId % 1) === 1)) {
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:88]: This one won't catch positive integers either.

File: express-notes-api/index.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+// id is not a +int or no cont prop in req body: 400 status code + JSON error obj
+// valid id but no cont prop: 404 status code + JSON error obj
+// valid id + cont prop: note at id is updated with content value + 200 status code
+
+app.put('/api/notes/:id', (req, res) => {
+  let changeId = Number(req.params.id);
+  const reqBody = req.body;
+  let exists;
+  let error;
+  if ((changeId > 0) && ((changeId % 1) === 1) && (reqBody !== undefined)) {
```
Comment (TimelikeClosure) [inline express-notes-api/index.js:123]: Neither will this.

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]

**Review by TimelikeClosure — APPROVED** [review]

**Pull request merged**